### ISSUE TITLE
[WAGON-381] Both Maven 2 and 3 fail to retrieve a <dependency> that is larger than Integer.MAX_VALUE

### ISSUE THREAD

**Issue by olamy** [olamy/maven-wagon#146]
**[Evgeny Goldin](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=genie)** opened **[WAGON-381](https://issues.apache.org/jira/browse/WAGON-381?redirect=false)** and commented

We have a **`*.tar`** file stored in corporate Maven repository, its size is **`2.5Gb`**. Trying to bring it with Maven (both 2 and 3) causes file of **`2147483647`** size to be downloaded to local maven repo, which is exactly 2\<sup>31\</sup>-1.

---

**Affects:** 2.2

**Attachments:**
- [1.png](https://issues.apache.org/jira/secure/attachment/12720682/1.png) (_31.86 kB_)
- [2.png](https://issues.apache.org/jira/secure/attachment/12720791/2.png) (_18.56 kB_)

1 votes, 7 watchers

**Comment by olamy** [comment]
**[Mark Struberg](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=struberg)** commented

which servlet engine is your corporate maven repository using? Not sure if this is a maven problem or a problem of the repo manager (sending the wrong size). We would need to test this with downloading from a plane httpd.

**Comment by olamy** [comment]
**[Evgeny Goldin](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=genie)** commented

Hi Mark,

We use Artifactory and downloading over HTTP works Ok with it. I'll see what headers are sent in response when Maven is making a request for \<dependency>.

**Comment by olamy** [comment]
**[Evgeny Goldin](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=genie)** commented

I tried to sniff the headers when dependency is sent but it's not that easy. IEInspector HTTP Analyzer crashes, WireShark freezes and I can't see the headers. But on one occasion before HTTP Analyzer crashed I saw the correct response length header was sent. I'll try to see if there are other ways to see the headers. May be you can also reproduce this problem and sniff the traffic with better tools or better machines than mine.

**Comment by olamy** [comment]
**[Benjamin Bentmann](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=bentmann)** commented

[URLConnection.getContentLength()](http://download.oracle.com/javase/1.5.0/docs/api/java/net/URLConnection.html#getContentLength%28%29) is of type `int`, this might explain the trouble with wagon-http-lightweight.

**Comment by olamy** [comment]
**[Evgeny Goldin](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=genie)** commented

Managed to grab headers again, before the sniffer crashed (see images attached). Yes, the correct length is sent. Was running Maven 2.2.1 and 3.0.2.

**Comment by olamy** [comment]
**[David McWhorter](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=davidm2005)** commented

I have the same issue, using the latest maven 3.0.4.  Downloads exactly 2^31-1 bytes and then stops.  Also using artifactory to serve the artifact and downloading the artifact with wget works without problems.

**Comment by olamy** [comment]
**[Olivier Lamy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=olamy)** commented

must be fixed with https://git-wip-us.apache.org/repos/asf/maven-wagon/repo?p=maven-wagon.git;a=commit;h=b8ca4254b1f5a828e7b1347c49b5f14bbb59e30b
You can try last maven 3.1 snapshot available here https://builds.apache.org/view/M-R/view/Maven/job/maven-3.0.x/
`@since` build #365

**Comment by olamy** [comment]
**[Ben Douglas](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=bennid)** commented

I just attempted the 3.1 snapshot (build 320) and my download stalled again and gave me a bad md5 warning:

```
bdouglas@gambit templates$ /opt/apache-maven-3.1-SNAPSHOT/bin/mvn --version
Apache Maven 3.1-SNAPSHOT (rNON-CANONICAL_2012-12-01_03-31_jenkins; 2012-11-30 22:31:54-0500)
Maven home: /opt/apache-maven-3.1-SNAPSHOT
Java version: 1.6.0_33, vendor: Sun Microsystems Inc.
Java home: /opt/jdk1.6.0_33/jre
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "3.6.11-1.fc17.x86_64", arch: "amd64", family: "unix"
bdouglas@gambit templates$
```

The checksum failing is when the artifact is only the Integer.MAX_VALUE size

```
[INFO] Configured Artifact: org.centos:CentOS-6.3-x86_64-bin-DVD1:1.0:iso
[DEBUG] Using connector WagonRepositoryConnector with priority 0 for http://xxxxx.com/artifactory/libs-release
Downloading: http://xxxxx.com/artifactory/libs-release/org/centos/CentOS-6.3-x86_64-bin-DVD1/1.0/CentOS-6.3-x86_64-bin-DVD1-1.0.iso
[WARNING] Checksum validation failed, expected eaa52f3d1ccf2df3b03e064fb0fa6168be28a4d4 but is 249cf446611db34ad8006bbdb99397c2a4bc201a for http://xxxxx.com/artifactory/libs-release/org/centos/CentOS-6.3-x86_64-bin-DVD1/1.0/CentOS-6.3-x86_64-bin-DVD1-1.0.iso
```

**Comment by olamy** [comment]
**[Olivier Lamy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=olamy)** commented

only checksum fail but artifact is correctly downloaded ?

**Comment by olamy** [comment]
**[Ben Douglas](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=bennid)** commented

No -- looking at the download information, it stopped downloading then checked the checksum, it failed, so it retried the download, and kept trying and only got as far at the ~1G each time, finally failing

**Comment by olamy** [comment]
**[Olivier Lamy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=olamy)** commented

ok I just need to find a way how to test that.
As adding such huge file in the sources tree for unit test is impossible.

**Comment by olamy** [comment]
**[Gomez Henri](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hgomez)** commented

Tested with latest Maven 3.1-SNAPSHOT, same problem :

```
Apache Maven 3.1-SNAPSHOT (ac64dd6bb6096e8dbd54e3520208e5737fb1c804; 2013-05-29 18:48:30-0700)
Maven home: /root/apache-maven-3.1-SNAPSHOT
Java version: 1.6.0_45, vendor: Sun Microsystems Inc.
Java home: /opt/mycorp/jvm/java-1.6.0-sun-x64/jre
Default locale: fr_FR, platform encoding: UTF-8
OS name: "linux", version: "3.4.33-2.24-default", arch: "amd64", family: "unix"
```

Artifact is 2.5Gb long

**Comment by olamy** [comment]
**[Olivier Lamy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=olamy)** commented

still weird for me.
I added a unit test https://git-wip-us.apache.org/repos/asf?p=maven-wagon.git;a=commit;h=05098b3f0f0378a7272f0d9491ef356889bfc67f
And those huge files (with Content-Length or Chunked) are correctly downloaded.

**Comment by olamy** [comment]
**[Benson Margulies](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=bmargulies)** commented

We're seeing the same thing here with Nexus as the repo.

**Comment by olamy** [comment]
**[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas)** commented

I was able to reproduce this with Wagon 2.4 (used in _all released Mavens since 3.0.5 and 3.1.x line_, even today's master uses Wagon 2.4). Also, I was able to test and verify it's fixed in Wagon 2.5-SNAPSHOT.

Setup:
Created 3G file, put it under my local Nexus (built from master). Using `curl -IL ...` verified that Nexus emits correct headers, especially that `Content-Length` header is correct:

```
[cstamas@Zaphod nexus-perftest]$ curl -IL http://localhost:8081/nexus/content/groups/public/org/test/test/1.0/test-1.0-data.dat
HTTP/1.1 200 OK
Date: Thu, 12 Sep 2013 09:29:11 GMT
Set-Cookie: rememberMe=deleteMe; Path=/nexus; Max-Age=0; Expires=Wed, 11-Sep-2013 09:29:11 GMT
Server: Nexus/2.7.0-SNAPSHOT
Accept-Ranges: bytes
ETag: "{SHA1{6e7f6dca8def40df0b21f58e11c1a41c3e000285}}"
Content-Type: application/octet-stream
Last-Modified: Thu, 12 Sep 2013 09:15:14 GMT
Content-Length: 3221225472
X-Content-Type-Options: nosniff

[cstamas@Zaphod nexus-perftest]$ 
```

Then, I took "vanilla" Maven 3.0.4 (that should be plagued with this issue, and is the first to use non-lightweight Wagon for HTTP) and built a sample project, noticed following lines in console proving the existence of bug:

```
...
Downloading: http://localhost:8081/nexus/content/groups/public/org/test/test/1.0/test-1.0-data.dat
...
[WARNING] Checksum validation failed, no checksums available from the repository for http://localhost:8081/nexus/content/groups/public/org/test/test/1.0/test-1.0-data.dat
Downloaded: http://localhost:8081/nexus/content/groups/public/org/test/test/1.0/test-1.0-data.dat (2097152 KB at 38750.0 KB/sec)
...
```

Notice the file size Maven 3.0.4 reported, _it's not 3G_. Verified in local repo, only 2G were downloaded, file was incomplete.

Then, I locally built Wagon 2.5-SNAPSHOT (from 574561f6d5d1721813c2ec8902aed9e28022854b), then built Maven from master (from 02d12423060361d5444817083c64adf276c3953e) where I bumped Wagon version to 2.5-SNAPSHOT, nuked local repository and rerun the sample project build:

```
...
Downloading: http://localhost:8081/nexus/content/groups/public/org/test/test/1.0/test-1.0-data.dat
[WARNING] Checksum validation failed, no checksums available from the repository for http://localhost:8081/nexus/content/groups/public/org/test/test/1.0/test-1.0-data.dat
Downloaded: http://localhost:8081/nexus/content/groups/public/org/test/test/1.0/test-1.0-data.dat (3145728 KB at 86418.7 KB/sec)
...
```

Notice, how the reported file size is correct. Also, verified in local storage by comparing size and using `openssl sha1` that the file is complete and correctly downloaded (no corruption of any kinds happened), the hashes matched.

Also, please note that all Maven versions since 3.0.5 release (even 3.1.0 and 3.1.1 and even current master of Maven) all of them use Wagon 2.4, hence testing with latest Maven builds would not help, as all of those use Wagon 2.4, unless you used a branch with upped Wagon version, which in Henry's case is not true, it's a build from master branch
https://github.com/apache/maven/commit/ac64dd6bb6096e8dbd54e3520208e5737fb1c804

In case of Ben, it's a Jenkins build, probably another master branch build.

Hence, I consider the bug fixed, and am closing this issue. Please reopen, if you find some concerns not addressed.

**Comment by olamy** [comment]
**[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas)** commented

Closing the issue as the bug reported here is fixed in Wagon 2.5
